src/services/formatting/rulesMap.ts

```diff
@@ -1,60 +1,59 @@
-///<reference path='references.ts' />
+/// <reference path="rules.ts" />
 
 /* @internal */
 namespace ts.formatting {
-    export class RulesMap {
-        public map: RulesBucket[];
-        public mapRowLength: number;
+    export function getFormatContext(options: FormatCodeSettings): formatting.FormatContext {
+        return { options, getRule: getRulesMap() };
+    }
 
-        constructor(rules: ReadonlyArray<Rule>) {
-            this.mapRowLength = SyntaxKind.LastToken + 1;
-            this.map = new Array<RulesBucket>(this.mapRowLength * this.mapRowLength);
+    let rulesMapCache: RulesMap | undefined;
 
-            // This array is used only during construction of the rulesbucket in the map
-            const rulesBucketConstructionStateList: RulesBucketConstructionState[] = new Array<RulesBucketConstructionState>(this.map.length);
-            for (const rule of rules) {
-                this.FillRule(rule, rulesBucketConstructionStateList);
-            }
+    function getRulesMap(): RulesMap {
+        if (rulesMapCache === undefined) {
+            rulesMapCache = createRulesMap(getAllRules());
         }
+        return rulesMapCache;
+    }
 
-        private GetRuleBucketIndex(row: number, column: number): number {
-            Debug.assert(row <= SyntaxKind.LastKeyword && column <= SyntaxKind.LastKeyword, "Must compute formatting context from tokens");
-            return (row * this.mapRowLength) + column;
-        }
-
-        private FillRule(rule: Rule, rulesBucketConstructionStateList: RulesBucketConstructionState[]): void {
-            const specificRule = rule.descriptor.leftTokenRange.isSpecific() && rule.descriptor.rightTokenRange.isSpecific();
-
-            rule.descriptor.leftTokenRange.GetTokens().forEach((left) => {
-                rule.descriptor.rightTokenRange.GetTokens().forEach((right) => {
-                    const rulesBucketIndex = this.GetRuleBucketIndex(left, right);
+    export type RulesMap = (context: FormattingContext) => Rule | undefined;

```

Comment:
we need a different name for this type now.. RulesMap seems wrong.

Reply:
RuleMapper maybe..

Reply:
How about `GetRule`?

Reply:
@mhegazy Bump

Reply:
GetRule seems like a function to me and not a type name..

Reply:
@DanielRosenwasser or @sandersn might have better naming ideas.